src/DataGrid.tsx

```diff
@@ -154,52 +175,83 @@ function DataGrid<R, K extends keyof R, SR>({
   cellNavigationMode = CellNavigationMode.NONE,
   editorPortalTarget = document.body,
   defaultFormatter = ValueFormatter,
-  columns,
+  columns: rawColumns,
   rows,
+  rowRenderer,
+  rowGroupRenderer,
+  summaryRows,
   selectedRows,
+  onRowClick,
+  onRowExpandToggle,
+  onSelectedCellChange,
+  onSelectedCellRangeChange,
   onSelectedRowsChange,
   ...props
 }: DataGridProps<R, K, SR>, ref: React.Ref<DataGridHandle>) {
-  const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
-  const [scrollLeft, setScrollLeft] = useState(0);
-  const [gridWidth, setGridWidth] = useState(0);
+  /**
+   * refs
+   * */
   const gridRef = useRef<HTMLDivElement>(null);
-  const headerRef = useRef<HTMLDivElement>(null);
-  const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
+  const lastSelectedRowIdx = useRef(-1);
 
-  const columnMetrics = useMemo(() => {
-    if (viewportWidth <= 0) return null;
+  /**
+   * states
+   */
+  const [eventBus] = useState(() => new EventBus());
+  const [gridWidth, setGridWidth] = useState(0);
+  const [scrollTop, setScrollTop] = useState(0);
+  const [scrollLeft, setScrollLeft] = useState(0);
+  const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
+
+  /**
+   * computed values
+   */
+  const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
 
+  const { columns, lastFrozenColumnIndex, totalColumnWidth } = useMemo(() => {
     return getColumnMetrics<R, SR>({
```

Comment:
Do we not need `if (viewportWidth <= 0) return null;` check anymore?

Reply:
It works fine without it, so I guess not, and it simplifies things a bit.